feat: Convert Select from Main into a compound component (#2309)
Workday/canvas-kit: modules/react/select/stories/Select.stories.mdx

Reviewer: I noticed some strange scrolling behavior when clicking on examples in the Storybook docs.

1. Load the Select docs in SB ([latest build](https://5d854c26ba934e0020f5e98a-kyqhsbtnco.chromatic.com/?path=/docs/components-inputs-select--basic) as of the writing of this comment)
2. Scroll to an example further down on the page, such as `With Icons`.
3. Click on the Select Input and select any item in the menu **other than the first one**. The menu will collapse and the selected item will be represented in the Select Input.
4. Click on the Select Input again. The page will scroll to the top.

Strangely, if your **first and only** selection is the first item in the menu, the scrolling will not occur. For example, if you select `Activity Stream` in the `With Icons` example, no scrolling will occur when you click on the Select Input again.

Author: interesting....

Author: I wonder if that has anything to do with virtualization, which I set to false in the model

Author: Looks like the dom element for the popup is still in the dom even after it is closed

### Basic Example

`Select` supports a
[dynamic API](/getting-started/for-developers/resources/collection-api/#dynamic-items) where you
pass an array of items via the `items` prop and provide a render function to display the items. The
items may be provided as an
[array of strings](/getting-started/for-developers/resources/collection-api/#array-of-strings) or an

> [array of objects](/getting-started/for-developers/resources/collection-api/#array-of-objects).

Reviewer: [Your original Select docs](https://5d854c26ba934e0020f5e98a-bdjnqbgofl.chromatic.com/?path=/docs/components-inputs-select--basic), for reference.

Originally, you had a separate `Registering Items` section in the Select MDX with examples. I decided to move that content into the `Basic Example` and link out to the `Array of Strings` and `Array of Objects` sections in the Collection MDX. It seemed like that would be sufficient instead of repeating that content here.

`Select` should be used in tandem with [Form Field](/components/inputs/form-field/) where the
`inputId` prop matches an `id` on the `Select.Input` to meet accessibility standards. This ensures
the `label` text from `FormField` is attached to the `Select.Input` and read out as a group for
voiceover.

<ExampleCodeBlock code={Basic} />

Our example uses [React state](<(https://react.dev/learn/state-a-components-memory)>) to track the
value of the `Select`.

### Hoisted Model

By default, `Select` will create and use its own model internally. Alternatively, you may configure
your own model with `useSelectModel` and pass it to `Select` via the `model` prop. This pattern is
referred to as
[hoisting the model](/getting-started/for-developers/resources/compound-components/#configuring-a-model)

> and provides direct access to its `state` and `events` outside of the `Select` component.

Reviewer: [Your original Select docs](https://5d854c26ba934e0020f5e98a-bdjnqbgofl.chromatic.com/?path=/docs/components-inputs-select--basic), for reference.

I updated the `Hoisted Model` section and modeled it after the [`Hoisted Model` section in Tabs](https://github.com/Workday/canvas-kit/blob/master/modules/react/tabs/stories/Tabs.stories.mdx#L59). I also updated the example itself to provide a use case for why you might want to hoist the model (example is contrived, but it should get the job done).

`disabled` is used to populate `nonInteractiveIds` and set `aria-disabled` for the disabled items.

<ExampleCodeBlock code={DisabledOptions} />

### With Icons

Use `Select.Item.Icon` to render an icon for a `Select.Item`. The `icon` prop for `Select.Item.Icon`
accepts [system icons](/assets/system-icons/) from `@workday/canvas-system-icons-web`.

In order to render the icon for the selected item in the `Select.Input`:

1. Obtain a reference to the `model` by registering your `items` with `useSelectModel`.
2. Get the selected item:
   `const selectedItem = model.navigation.getItem(model.state.selectedIds[0], model)`
3. Pass the icon for the selected item to the input:
   `<Select.Input inputStartIcon={selectedItem.value.icon}>`

<ExampleCodeBlock code={WithIcons} />

Note that `Select.Input` will only render an icon if an item is selected.

### Grow

Set the `grow` prop of the wrapping `FormField` to `true` to configure the `Select.Input` to expand
to the width of its container.

<ExampleCodeBlock code={Grow} />

### Menu Height

`Select.Card` has a default maximum height of `300px` to restrict the height of the dropdown menu.
Set its `maxHeight` prop to override this value.

> <ExampleCodeBlock code={MenuHeight} />

Reviewer: I renamed the `Scrollable` section and example to `Menu Height` to better reflect its content related to `maxHeight`.

### Error States

Set the `error` prop of the wrapping `FormField` to `FormField.ErrorType.Alert` or
`FormField.ErrorType.Error` to set the `Select` to the alert or error state, respectively. You will
also need to set the `hintId` and `hintText` props on the `FormField` to meet accessibility
standards. You must set an `id` attribute on the `Select.Input` element that matches the value of
`inputId` set on the `FormField` element. These attributes ensure that the alert message is
associated to the `Select` and read out by voiceover.

**Note: The Select container component, `Select`, must wrap `FormField` to ensure `Select.Input` is
styled correctly.**

```tsx
<Select items={options}>
  <FormField label="Contact" inputId="contact-id-formfield">
    <Select.Input id="contact-id-formfield" />
    ...
  </FormField>
</Select>
```

#### Alert

Use the alert state when a selection is valid but there is additional information.

<ExampleCodeBlock code={Alert} />

#### Error

Use the error state when the selection is no longer valid.

<ExampleCodeBlock code={Error} />

## Component API

<SymbolDoc name="Select" fileName="/react/" />

## Specifications

> <Specifications file="Select.spec.ts" name="Select" />

Reviewer: We should add the Specifications back in, no?

value of the `Select`.

### Label Position

Set the `labelPosition` prop of the wrapping `FormField` to designate the position of the label
relative to the `Select`. `labelPosition` accepts the following values:

- `FormField.LabelPosition.Top` (Default)
- `FormField.LabelPosition.Left`

<ExampleCodeBlock code={LabelPosition} />

### Required

Set the `required` prop of the wrapping `FormField` to `true` to indicate that the field is
required. Labels for required fields are suffixed by a red asterisk.

<ExampleCodeBlock code={Required} />

### Disabled

Set the `disabled` prop of `Select.Input` to prevent users from interacting with it.

<ExampleCodeBlock code={Disabled} />

### Disabled Items

To disable items and prevent them from being interactive, you must set two properties.

Set `nonInteractiveIds` to an array of `id`s that you want to disable interaction as well as
searchability when typing into the `Select`.

You will also need to set the aria attribute `aria-disabled` on the `Select.Item` if the item is
disabled, otherwise set the attribute to `undefined`. `disabled` in this case is a property on the

> item's object. This ensures that the item is styled as disabled.

Reviewer: GH isn't allowing me to include a multi-line suggestion for the entire `Disabled Items` section, so I'm pasting it here:

```
### Disabled Items

In order to disable items and prevent users from interacting with them:

1. Set the `nonInteractiveIds` prop of `Select` to an array of disabled item `id`s. This will
   disable interaction for those items and exclude them from type-ahead.

2. Set the `aria-disabled` attribute of all disabled `Select.Item`s to `true`. This ensures the
   items are styled as disabled.

The following example assigns a `disabled` property to the disabled items in the `items` array.
`disabled` is used to populate `nonInteractiveIds` and set `aria-disabled` for the disabled items.
```

I did a few things here:

1. Updated language for consistency.
2. Replaced "searchability" with "type-ahead." I lifted the "type-ahead" nomenclature from the [WAI Listbox](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/) page.
3. Moved the bit about the `disabled` property in `items` to the very end.

Author: Nice, i like the clear numbered steps, and the separation of the info of objects.

to the width of its container.

<ExampleCodeBlock code={Grow} />

### Scrollable

Set `maxHeight` prop on the `Select.Card` to determine the height of the dropdown menu. This will
allow the menu to scroll if there are enough items.

<ExampleCodeBlock code={Scrollable} />

### Ref Forwarding

`Select.Input` supports [ref forwarding](https://reactjs.org/docs/forwarding-refs.html). It will
forward `ref` to its underlying `<input type="text" role="combobox">` element.

<ExampleCodeBlock code={RefForwarding} />

### Hoisted Model

`<Select>` supports a
[dynamic API](/getting-started/for-developers/resources/collection-api/#dynamic-items) where you
pass an array of `items` in the `model` config and provide a render function to display the items.
It's important to set up the object in your items array in a way for the
[collection system](/getting-started/for-developers/resources/collection-api/) to understand.

#### Array of strings

If you have an array of `strings`, the collection system will handle asigning an `id`, `index` and
`textValue` to each item in the array. This ensures that both type ahead and keyboard navigation

> work correclty.

Reviewer: ```suggestion
work correctly.
```

relative to the `Select`. `labelPosition` accepts the following values:

- `FormField.LabelPosition.Top` (Default)
- `FormField.LabelPosition.Left`

<ExampleCodeBlock code={LabelPosition} />

### Required

Set the `required` prop of the wrapping `FormField` to `true` to indicate that the field is
required. Labels for required fields are suffixed by a red asterisk.

<ExampleCodeBlock code={Required} />

### Disabled

Set the `disabled` prop of `Select.Input` to prevent users from interacting with it.

<ExampleCodeBlock code={Disabled} />

### Disabled Items

In order to disable items and prevent users from interacting with them:

1. Set the `nonInteractiveIds` prop of `Select` to an array of disabled item `id`s. This will
   disable interaction for those items and exclude them from type-ahead.

2. Set the `aria-disabled` attribute of all disabled `Select.Item`s to `true`. This ensures the
   items are styled as disabled.

The following example assigns a `disabled` property to the disabled items in the `items` array.
`disabled` is used to populate `nonInteractiveIds` and set `aria-disabled` for the disabled items.

<ExampleCodeBlock code={DisabledOptions} />

> ### With Icons
> 
> Add `Select.Item.Icon` as a sub component of `Select.Item` if you want to render an icon for each
> item. The `Select.Item.Icon` takes in an `icon` prop of type `CanvasSystemIcon`. You can import
> icons from `@workday/canvas-system-icons-web`.
> 
> If you want to render the selected item's icon in the `Select.Input` add the `inputStartIcon` prop.
> 
> To get the current selected item's icon:
> 
> - Register your `items` with the `useSelectModel`
> - Get current item by
>   `const currentItem = model.navigation.getItem(model.state.selectedIds[0], model)`
> - Pass the value to the input in the following way:
>   `<Select.Input inputStartIcon={currentItem.value.icon} id="form-field-input-id" />`
> 
> **Notes: The `Select.Input` will only render an icon when there is an item selected.**
> 
> <ExampleCodeBlock code={WithIcons} />

Reviewer: A few updates for the `With Icons` example:

1. Cleaned up and updated language for consistency.
2. Linked to the `/assets/system-icons/` route. This content isn't great in SB, but it's better on the Canvas Site.
3. Moved the note about `Select.Input` to the end after the example and removed the bold since it didn't seem necessary to emphasize this.

```suggestion
### With Icons

Use `Select.Item.Icon` to render an icon for a `Select.Item`. The `icon` prop for `Select.Item.Icon` accepts
[system icons](/assets/system-icons/) from `@workday/canvas-system-icons-web`.

In order to render the icon for the selected item in the `Select.Input`:

1. Obtain a reference to the `model` by registering your `items` with `useSelectModel`.
4. Get the selected item: `const selectedItem = model.navigation.getItem(model.state.selectedIds[0], model)`
5. Pass the icon for the selected item to the input: `<Select.Input inputStartIcon={selectedItem.value.icon}>`

<ExampleCodeBlock code={WithIcons} />

Note that `Select.Input` will only render an icon if an item is selected.
```

					);
				}}
			</Select.List>
		</Select.Card>
	</Select.Popper>
</Select>
Selected Value: {model.state.selectedIds[0]}
</>
```

<ExampleCodeBlock code={HoistedModel} />

### Error States

Set the error prop of the wrapping `FormField` to `FormField.ErrorType.Alert` or
`FormField.ErrorType.Error` to set the `<Select>` to the alert or error state, respectively. You
will also need to set the `hintId` and `hintText` props on the `FormField` to meet accessibility
standards. You must set an `id` attribute on the `Select.Input` element that matches the value of
`inputId` set on the `FormField` element. These attributes ensure that the alert message is
associated to the `Select` and read out by voiceover.

**Note: The Select container component, `Select` must wrap the `FormField` element to ensure correct
styling on the `Select.Input` element.**

```tsx
<Select items={options}>
	<FormField label="Contact" inputId="contact-id-formfield">
		<Select.Input id="contact-id-formfield" />
		... select sub components
	</FormfField>
</Select>
```

#### Alert

Use the alert state when a selection is valid but there is additional information.

<ExampleCodeBlock code={Alert} />

#### Error

> Use the error state when the selection is no longer valid.
> 
> > **Note**
> >
> > The blade-old package has been moved to it's own repo.
> > [github.com/razorpay/blade-old](https://github.com/razorpay/blade-old) (Only accessible to Razorpay employees)
> >
> > It is under **maintenance** and it won't have any major releases.

Reviewer: This is a strange note. An inadvertent copy-and-paste?

```suggestion
Use the error state when the selection is no longer valid.
```

Author: lol yes, I was testing out creating a note, sorry!

### Grow

Set the `grow` prop of the wrapping `FormField` to `true` to configure the `Select.Input` to expand
to the width of its container.

<ExampleCodeBlock code={Grow} />

### Scrollable

Set `maxHeight` prop on the `Select.Card` to determine the height of the dropdown menu. This will
allow the menu to scroll if there are enough items.

<ExampleCodeBlock code={Scrollable} />

### Ref Forwarding

`Select.Input` supports [ref forwarding](https://reactjs.org/docs/forwarding-refs.html). It will
forward `ref` to its underlying `<input type="text" role="combobox">` element.

<ExampleCodeBlock code={RefForwarding} />

### Hoisted Model

`<Select>` supports a
[dynamic API](/getting-started/for-developers/resources/collection-api/#dynamic-items) where you
pass an array of `items` in the `model` config and provide a render function to display the items.
It's important to set up the object in your items array in a way for the
[collection system](/getting-started/for-developers/resources/collection-api/) to understand.

<ExampleCodeBlock code={HoistedModel} />

### Registering Items

You can register items in two way.

1. An Array of Strings

> 2. An Array objects

Author: fix
